pkg/controller/nodefabricnetworkattachments.go

```diff
@@ -532,6 +535,71 @@ func (cont *AciController) applyNodeFabNetAttObjLocked(vlans []int, addNet *Addi
 	}
 }
 
+func (cont *AciController) addContractsToEpg(netFabConfig *fabattv1.NetworkFabricConfiguration) apicapi.ApicSlice {
+	// TODO - handle case -> !cont.config.AciUseGlobalScopeVlan
+	var apicSlice apicapi.ApicSlice
+	if netFabConfig != nil {
+		appProfile := "netop-" + cont.config.AciPolicyTenant
+		tenantName := netFabConfig.Spec.VlanRefs[0].Epg.Tenant
+
+		if tenantName != cont.config.AciPolicyTenant {
+			cont.log.Warningf("Different tenant '%s' provided in NetworkFabricConfiguration Resource, Ignoring it and using AciPolicyTenant '%s'", tenantName, cont.config.AciPolicyTenant)
+			tenantName = cont.config.AciPolicyTenant
+		}
+		commonTenantDn := "uni/tn-common"
+		epgObj := netFabConfig.Spec.VlanRefs[0].Epg
+		bdObj := epgObj.BD
+		vrfObj := bdObj.Vrf
+
+		// create BD
```

Comment:
Do not create BD and EPG here. This is part of the function createNodeFabNetAttBd. Conditionally you have to either use the default prefix or the names in the CR. Contract should be a separate object with ref counts, do not create per nodefabnetatt object. This whole patch should be refactored.